crates/uv-resolver/src/resolver/batch_prefetch.rs

```diff
@@ -0,0 +1,172 @@
+}
+
+impl BatchPrefetcher {
+    /// Prefetch a large number of versions if we already unsuccessfully tried many versions.
+    pub(crate) async fn prefetch_batches(
+        &mut self,
+        next: &PubGrubPackage,
+        version: &Version,
+        current_range: &Range<Version>,
+        request_sink: &Sender<Request>,
+        index: &InMemoryIndex,
+        selector: &CandidateSelector,
+    ) -> anyhow::Result<(), ResolveError> {
+        let PubGrubPackage::Package(package_name, _, _) = &next else {
+            return Ok(());
+        };
+
+        let (num_tried, do_prefetch) = self.should_prefetch(next);
+        if !do_prefetch {
+            return Ok(());
+        }
+        let total_prefetch = min(num_tried, 50);
+
+        // This is immediate, we already fetched the version map.
+        let versions_response = index
+            .packages
+            .wait(package_name)
+            .await
+            .ok_or(ResolveError::Unregistered)?;
+
+        let VersionsResponse::Found(ref version_map) = *versions_response else {
+            return Ok(());
+        };
+
+        let mut phase = BatchPrefetchStrategy::Compatible {
+            compatible: current_range.clone(),
+            previous: version.clone(),
+        };
+        let mut i = 0;
+        while i < total_prefetch {
+            i += 1;
+            let candidate = match phase {
+                BatchPrefetchStrategy::Compatible {
+                    compatible,
+                    previous,
+                } => {
+                    if let Some(candidate) =
+                        selector.select_no_preference(package_name, &compatible, version_map)
+                    {
+                        let compatible = compatible.intersection(
+                            &Range::singleton(candidate.version().clone()).complement(),
+                        );
+                        phase = BatchPrefetchStrategy::Compatible {
+                            compatible,
+                            previous: candidate.version().clone(),
+                        };
+                        candidate
+                    } else {
+                        // We exhausted the compatible version, switch to ignoring the existing
+                        // constraints on the package and instead going through versions in order.
+                        phase = BatchPrefetchStrategy::InOrder { previous };
+                        continue;
+                    }
+                }
+                BatchPrefetchStrategy::InOrder { previous } => {
+                    let range = if selector.use_highest_version(package_name) {
+                        Range::strictly_lower_than(previous)
+                    } else {
+                        Range::strictly_higher_than(previous)
+                    };
+                    if let Some(candidate) =
+                        selector.select_no_preference(package_name, &range, version_map)
```

Comment:
I feel like this risks being quadratic, because we iterate over `total_prefetch`, and then `select_no_preference` is also linear, right? Is there any way to improve that?

Reply:
I did an attempt turning `select_candidate` into an iterator but then looked at the profiles for urllib3/boto3/botocore and i think it's not worth it.